Feature: session loop sorting and session activity format variable

mgrant0 opened the pull request:
Adds sorting flags to the S: format.  

Adds `session_alert` format which returns `#` for activity, `!` for bell, and `~` for silence if one of the windows in has those alerts.

Adds `session_bell_flag` format which is true if one of the windows had rung the bell in one of the windows in a session.

mgrant0 commented:
Should I add the sorting options to `S:` and perhaps also `P:`?

ThomasAdam commented:
> I left out `format_cb_session_alert` for now because if multiple windows have the same alert it seems like it will include them multiple times? Is that what you want?

Indeed -- I hope not.  This would really irritate me unless it can be predicated by checking if the window is linked...

nicm commented:
Looks good, can you add them to tmux.1 as well? There is a list of format variables.

mgrant0 commented:
I added `/i`, `/n`, `/t`, `/r` sorting to `W:` and `/r` sorting to `P:` because there's no pane name or pane last activity time.

I had a small issue with `W:` which caused me a bit of drama.  When you sort it other that by index ascending (the default), the callback `format_cb_window_end_flag` returns the last window by index which is wrong.  `OPTIONS_TABLE_STATUS_FORMAT1` uses `window_end_flag` to put a separator in between windows.

I fixed this by adding a new flag `loop_last_flag` which returns `1` on the last item in the loop.  (or feel free to rename it).  This, then, can be used in all the new and improved status lines.  To get this to work, I add a new FORMAT_LAST flag into the ft->flags.  I modified the first status line to use this new `loop_last_flag` instead of `window_end_flag` which I did not want to modify in case someone already uses that.

I also added the sorting flags to L: for clients, but since a client doesn't have a numeric id, I just used the order clients came out of the TAILQ.

nicm commented:
I mean that `session_bell_flag` does not go after `window_bell_flag` in `format_table`, it needs to go between `session_attached_list` and `session_created`, or `bsearch` will get confused.

There is `window_activity_flag` and `window_silence_flag`, the session versions of those can work the same way as `session_bell_flag` does for `window_bell_flag`?

mgrant0 commented:
Got it.  I think that was the only one out of order.  I did this to check it:
```
grep  '{ "' format.c > 1
sort 1 > 2
diff 1 2
```
I get no differences now.

mgrant0 commented:
I've added these so far:
- format_cb_session_silence_flag
- format_cb_session_activity_flag
- format_cb_session_active
- format_cb_last_session_index
- format_cb_session_index

What others might be useful?  Zoomed (as in, at least one window zoomed within a session) does not seem terribly useful.  Similar with a marked pane, does not seem useful to me in context of a session.  I was not going to implement a session stack to track last session.

I just need to clean things up, fix the ordering in the file and try to make sure they work before pushing these.

If there's something more you want here, let me know.

mgrant0 commented:
Let me know what else you want here.

nicm commented:
Looks good but:

- There is no such thing as session index, only session ID, so please get rid of `session_index` and `last_session_index`.

- `qsort_r` is not portable, so if you want to use it you will have to add a `compat` version.

nicm commented:
Since tmux is single threaded, a `compat` version of `qsort_r` should be trivial - it can call `qsort` with a global to store the context pointer.

nicm commented:
No, if OpenBSD had it then compat would be better but since it doesn't it's better to just stick with qsort so we don't have a difference between OpenBSD and portable.

mgrant0 commented:
Ok hopefully those commits address your comments.  Thanks.

naruaway commented on format.c:
```diff
@@ -4356,6 +4477,39 @@ format_window_name(struct format_expand_state *es, const char *fmt)
+
+	switch (sc.field) {
+	case FORMAT_LOOP_BY_INDEX:
+		result = wa->id - wb->id;
```
Due to this, windows are sorted by id by default, not by index, on the status bar

naruaway replied:
Fixing this in https://github.com/tmux/tmux/pull/4551
